src/charm.py

```diff
@@ -0,0 +1,314 @@
+
+    def _validate_config_params(self):
+        """Validate that configuration is valid.
+
+        Raises:
+            ValueError: in case when invalid charm-funcion is entered
+        """
+        charm_function = self.model.config["charm-function"]
+        if charm_function not in VALID_CHARM_FUNCTIONS:
+            raise ValueError(
+                f"config: invalid charm function {charm_function!r}"
+            )
+        default_role = self.config["self-registration-role"]
+        if default_role not in VALID_ROLES:
+            raise ValueError(
+                f"config: invalid self-registration role {default_role!r}"
+            )
+
+    def _create_env(self):
+        """Create state values from config to be used as environment variables.
+
+        Returns:
+            env: dictionary of environment variables
+        """
+        superset_secret = self.config.get(
+            "superset-secret-key"
+        ) or generate_random_string(32)
+        charm_function = self.config["charm-function"]
+        admin_email = self.config["oauth-admin-email"]
+        env = {
+            "SUPERSET_SECRET_KEY": superset_secret,
+            "ADMIN_PASSWORD": self.config["admin-password"],
+            "CHARM_FUNCTION": charm_function,
+            "SQL_ALCHEMY_URI": self._state.sql_alchemy_uri,
+            "REDIS_HOST": self._state.redis_host,
+            "REDIS_PORT": self._state.redis_port,
+            "ALERTS_ATTACH_REPORTS": self.config["alerts-attach-reports"],
+            "DASHBOARD_CROSS_FILTERS": self.config["dashboard-cross-filters"],
+            "DASHBOARD_RBAC": self.config["dashboard-rbac"],
+            "EMBEDDABLE_CHARTS": self.config["embeddable-charts"],
+            "SCHEDULED_QUERIES": self.config["scheduled-queries"],
+            "ESTIMATE_QUERY_COST": self.config["estimate-query-cost"],
+            "ENABLE_TEMPLATE_PROCESSING": self.config[
+                "enable-template-processing"
+            ],
+            "ALERT_REPORTS": self.config["alert-reports"],
+            "SQLALCHEMY_POOL_SIZE": self.config["sqlalchemy-pool-size"],
+            "SQLALCHEMY_POOL_TIMEOUT": self.config["sqlalchemy-pool-timeout"],
+            "SQLALCHEMY_MAX_OVERFLOW": self.config["sqlalchemy-max-overflow"],
+            "GOOGLE_KEY": self.config.get("google-client-id"),
+            "GOOGLE_SECRET": self.config.get("google-client-secret"),
+            "OAUTH_DOMAIN": self.config.get("oauth-domain"),
+            "OAUTH_ADMIN_EMAIL": admin_email,
+            "SELF_REGISTRATION_ROLE": self.config["self-registration-role"],
+        }
+        return env
+
+    def _update(self, event):
+        """Update the application server configuration and replan its execution.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        container = self.unit.get_container(self.name)
+        if not container.can_connect():
+            event.defer()
```

Comment:
This `defer()` should not be required, as the `PebbleReady` event will eventually call this method.